internal/kafka/internal/workers/kafka_mgrs/accepted_kafkas_mgr.go

```diff
@@ -78,35 +78,86 @@ func (k *AcceptedKafkaManager) Reconcile() []error {
 }
 
 func (k *AcceptedKafkaManager) reconcileAcceptedKafka(kafka *dbapi.KafkaRequest) error {
-	cluster, e := k.clusterService.FindClusterByID(kafka.ClusterID)
-	if cluster == nil || e != nil {
-		return errors.Wrapf(e, "failed to find cluster with '%s' for kafka request '%s'", kafka.ClusterID, kafka.ID)
-	}
-
-	latestReadyStrimziVersion, err := cluster.GetLatestAvailableAndReadyStrimziVersion()
-	if err != nil || latestReadyStrimziVersion == nil {
-		// Strimzi version may not be available at the start (i.e. during upgrade of Strimzi operator).
-		// We need to allow the reconciler to retry getting and setting of the desired strimzi version for a Kafka request
-		// until the max retry duration is reached before updating its status to 'failed'.
-		durationSinceCreation := time.Since(kafka.CreatedAt)
-		if durationSinceCreation < constants2.AcceptedKafkaMaxRetryDuration {
-			glog.V(10).Infof("No available and ready strimzi version found for Kafka '%s' in Cluster ID '%s'", kafka.ID, kafka.ClusterID)
-			return nil
-		}
-		kafka.Status = constants2.KafkaRequestStatusFailed.String()
+	var cluster *api.Cluster
+	kafkaAlreadyAssignedInADataPlaneCluster := kafka.ClusterID != ""
+	if !kafkaAlreadyAssignedInADataPlaneCluster {
+		assignedCluster, err := k.clusterPlacementStrategy.FindCluster(kafka)
 		if err != nil {
-			err = errors.Wrapf(err, "failed to get desired Strimzi version %s", kafka.ID)
-		} else {
-			err = errors.New(fmt.Sprintf("failed to get desired Strimzi version %s", kafka.ID))
+			return errors.Wrapf(err, "failed to find cluster for kafka request %s", kafka.ID)
+		}
+		if assignedCluster == nil {
+			return k.markTheUnassignedKafkaAsFailedOrAllowRetryClusterPlacementReconciliation(kafka)
 		}
-		kafka.FailedReason = err.Error()
-		if err2 := k.kafkaService.Update(kafka); err2 != nil {
-			return errors.Wrapf(err2, "failed to update failed kafka %s", kafka.ID)
+		kafka.ClusterID = assignedCluster.ClusterID
+		cluster = assignedCluster
+	} else {
+		foundCluster, e := k.clusterService.FindClusterByID(kafka.ClusterID)
+		if foundCluster == nil || e != nil {
+			return errors.Wrapf(e, "failed to find cluster with '%s' for kafka request '%s'", kafka.ClusterID, kafka.ID)
 		}
-		return err
+		cluster = foundCluster
 	}
-	kafka.DesiredStrimziVersion = latestReadyStrimziVersion.Version
 
+	latestReadyStrimziVersion, err := cluster.GetLatestAvailableAndReadyStrimziVersion()
+	if err != nil {
+		logger.Logger.Error(err)
+	}
+
+	noLatestReadyStrimziVersionFound := err != nil || latestReadyStrimziVersion == nil
```

Comment:
If an error happens I think we can just return the error instead of calling `markTheAssignedKafkaAsFailedOrAllowRetryStrimziVersionPickingReconciliation`. The same happens when trying to get the ibp version or the kafka version.

Reply:
Thanks. That makes sense, I've updated the code. I think for an error to occur here, it's mostly due to json parsing issues which we may not encounter at this stage as we only store valid json in the database.